Home page, structure and formatting improvements (#54)
nhsbsa-data-analytics/nhsbsa-analytical-code-assurance-playbook: guidance/06-documented/index.qmd

---

> title: "Documentation"

Reviewer: similar comment to other pages - page is fine but feels like we could have more here. particularly on READMEs and how to write good ones and what docstrings/other code documentation are etc.

Author: Captured by #58

description: "Documentation explains what your analytical project does, why decisions were made, and how others can use and reproduce your work."
---

> ## Suggested reading
>
> * [AF Duck book: Code documentation][duck-book-code-doc]
> * [AF Duck book: Project documentation][duck-book-project-doc]
> * [NHSBSA DDaT playbook: Licensing software or code][ddat-playbook-licenses]
> * [NHSBSA DDaT playbook: Standard code repository files][ddat-playbook-dev-doc]

## At a glance
Documentation is written information that explains what your analytical project does, why decisions were made, and how others can use and reproduce your work.

> This guide outlines the benefits of documentation and how to create and maintain quality documentation within the NHSBSA.

Reviewer: ```suggestion
This guide outlines the benefits of quality documentation and how to create and maintain it within the NHSBSA.
```
a few repeated words that interrupted flow

## What is documentation?
Documentation is written information that explains what your analytical project does, why decisions were made, and how others can use and reproduce your work.

> We should document:

Reviewer: ```suggestion
You should document:
```

* a project's purpose and scope
* how to setup the project and install any dependencies
* how to configure, use and run the code, including analytical pipelines, dashboards or webapps
* outputs from running the code and how to interpret them
* assumptions and limitations, how they are implemented and how they affect results
* key decisions and changes across a project lifecycle
* quality assurance logs 

## Why should we document our code?

Good documentation:

* Helps the intended audience understand the project using suitable language.
* Makes the project accessible to a wider audience and improves transparency.
* Helps to ensure the project is reusable and the results are reproducible.
* Supports knowledge sharing and collaboration between team members.
* Makes it easier for others to [review][peer-review] the code and project as a whole, including spotting potential issues or improvements.
* Reduces the risk of knowledge loss when team members change.

> * Provides an audit trail and evidence that sufficient [quality assurance][testing] has taken place.

Reviewer: ```suggestion
* helps the intended audience understand the project using suitable language
* makes the project accessible to a wider audience and improves transparency
* helps to ensure the project is reusable and the results are reproducible
* supports knowledge sharing and collaboration between team members
* makes it easier for others to [review][peer-review] the code and project as a whole, including spotting potential issues or improvements
* reduces the risk of knowledge loss when team members change
* provides an audit trail and evidence that sufficient [quality assurance][testing] has taken place
```
remove capitalisation and punctuation

## How do we create and maintain quality documentation?

Treat documentation as an ongoing activity rather than a final task. The workflow below maps documentation to natural points in a project lifecycle:

1. At project outset set up the [standard repository files][ddat-playbook-dev-doc] using common filenames and formats, and agree who is responsible for each part of the documentation. Allocate time for documentation in your estimates.

> 2. As you code write comments to explain *why* decisions were made, not *what* the code does. Use standard docstrings for functions and modules, and consider tools that generate reference documentation automatically from these. Use simple language appropriate for your audience, adding extra documentation where the project is complex (e.g. design documents, accessibility statements, or user help).

Reviewer: ```suggestion
2. As you code write comments to explain **why** decisions were made, not **what** the code does. Use standard docstrings for functions and modules, and consider tools that generate reference documentation automatically from these. Use simple language appropriate for your audience, adding extra documentation where the project is complex. For example, design documents, accessibility statements, or user help.
```
tweaks for accessibility - remove italicisation and expand e.g.
